### ISSUE TITLE
Fast weight update within the chunk

### ISSUE THREAD

**Issue by Forrest-110** [a1600012888/LaCT#15]
Hi authors, thank you for your incredible work!

I noticed a difference between your implementation and the official TTT implementation regarding fast weight updates. In the [official TTT implementation](https://github.com/test-time-training/ttt-lm-pytorch/blob/cd831db10c8c9a0f6340f02da5613316a8a92b67/ttt.py#L1003-L1010), the fast weights are updated within the chunk—meaning each token in the same chunk queries from a different (and progressively updated) fast weight.

However, in your [LLM implementation](https://github.com/a1600012888/LaCT/blob/c89091740f462571e190fdb44af4b130b740893f/lact_llm/lact_model/ttt_operation.py#L164-L169), it appears that the fast weight is shared across the entire chunk and only updated once per chunk.

Does this imply that, in your implementation, if all other attention layers are removed and only the TTT layer remains, tokens within the same chunk would not be able to see each other?

Thank you for your time and attention!

**Comment by a1600012888** [comment]
Hi Yifei,

- Difference with the original TTT.
You are correct.  If you think everything from the viewpoint of the equivalant attention mask.  The original TTT does per-token causal mask. 

Here if you do chunk-update, then apply the updated fast weight to the same chunk of tokens, you are doing a bidirectional attention mask within the chunk, i.e. the tokens within each chunk can see each other (Like figure 3(a) or our paper)

Then you do apply first, then do update. Then it's like figure 3(c), and tokens within each chunk are not visible to each other. And that's what we do in the language model case.
